CONTRIBUTING.md

```diff
@@ -10,8 +10,20 @@ contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additio
+The 'Create Cluster' and 'Configure from Cluster' commands rely on provider-specific tools or APIs for interacting with the cloud or cluster.  We have a basic extension point for implementing new cloud providers.  **This is currently experimental and we welcome feedback.**
 
-At the moment, the extension provides a generic (albeit fiddly) mechanism for managing a series of pages in a UI (example: displaying a list of accounts for the user to choose from), invoking tools or APIs for populating state to be used in constructing those pages (example: invoking the Azure CLI to get a list of subscriptions), and flowing choices and state between those pages (example: returning the selected account so that another tool can list the clusters in that account).  However, the cloud-specific pages and logic have to be built using artisanal TypeScript and built into this extension via pull request - we don't provide a nice way for you to externalise different providers.
+To implement a provider for the 'Create Cluster' and/or 'Configure from Cluster' commands, your extension needs to do the following:
 
-We'd be _really really happy_ to work with anyone wanting to implement providers for e.g. minikube, GKE, EKS, etc., both to help you understand how the existing implementation works, and to collaborate on simplifying and stabilising an extension mechanism.
+* Set itself for automatic activation (the `*` activation event).  We know this isn't great and would love to discuss a better approach.
+* Start a HTTP server listening on a port of your choice (see `port` below).  This HTTP server will serve up the HTML pages by which the user chooses a cluster or specifies creation settings.
+* Get the VS Code Kubernetes extension's API object via the `vscode.extensions.getExtension<T>` and `Extension<T>.activate()` functions.
+* The API object has a field named `clusterProviderRegistry`.  This is an object you can use to register the type(s) of cluster you support.  Call the registry's `register` method, passing an object with the following fields:
+  * `id` (string): an identifier for the type of cluster.  This needs to be unique across all providers, as we use it to dispatch the user's selected cluster type to the appropriate handler.  E.g. for Azure Kubernetes Service we use `aks`.
+  * `displayName` (string): the display name for the type of cluster.  E.g. for Azure Kubernetes Service we use (wait for it) `Azure Kubernetes Service`.
+  * `port` (number): the port on which your provider will serve the requisite configuration pages.  We reserve port 44010 for the internal 'select cluster type' page provider.  Also, the Azure provider uses port 44011, so please don't use that either.
```

Comment:
I wonder if we'll need to reserve these ports, since the cluster registry is the entry point for creation and configuration of clusters. I'm thinking that If we find an unbound port via node, we can bind to it, and pass it back to the clusterRegistry.register() function, eliminating the "reserved" port thing altogether. A quick npm search reveals this module, which might be useful: https://www.npmjs.com/package/get-port

Reply:
We already have portfinder in the project from the debugging work, but I tried using it to replace the literal ports and it didn't work (the second server was assigned the same port as the first, even though the first one was already listening).  @testforstephen any advice on the successful use of portfinder?  If I can't get it working, I'll leave it as fixed ports, as it will be a non-breaking change to make them dynamic later.

Reply:
Brilliant!  Thank you!